QuickFIXn/DataDictionary/DataDictionary.cs

```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
-                    {
-                        ddmap.ReqFields.Add(fld.Tag);
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
+                        DDField fld = FieldsByName[nameAttribute];
+                        if (required && (componentRequired == null || componentRequired.Value == true))
+                        {
+                            ddmap.ReqFields.Add(fld.Tag);
+                        }
+                        if (!ddmap.IsField(fld.Tag))
+                        {
+                            ddmap.Fields.Add(fld.Tag, fld);
+                        }
 
-                    // if first field in group, make it the DELIM
-                    if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
-                    {
-                        ((DDGrp)ddmap).Delim = fld.Tag;
-                    }
-                }
-                else if (childNode.Name == "group")
-                {
-                    DDField fld = FieldsByName[fieldName];
-                    DDGrp grp = new DDGrp();
-                    XmlAttribute req = childNode.Attributes["required"];
-                    if (req != null && req.Value == "Y"
-                        && (componentRequired == null || componentRequired.Value == true))
-                    {
-                        ddmap.ReqFields.Add(fld.Tag);
-                        grp.Required = true;
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
-                    grp.NumFld = fld.Tag;
-                    ParseMsgEl(childNode, grp);
-                    ddmap.Groups.Add(fld.Tag, grp);
+                        // if this is in a group whose delim is unset, then this must be the delim (i.e. first field)
+                        if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
+                        {
+                            ((DDGrp)ddmap).Delim = fld.Tag;
+                        }
+                        break;
 
-                    // if first field in group, make it the DELIM
-                    if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
-                    {
-                        ((DDGrp)ddmap).Delim = fld.Tag;
-                    }
-                }
-                else if (childNode.Name == "component")
-                {
-                    XmlNode compNode = RootDoc.SelectSingleNode("//components/component[@name='" + fieldName + "']");
-                    XmlAttribute req = childNode.Attributes["required"];
-                    bool? compRequired = (req != null && req.Value == "Y");
-                    ParseMsgEl(compNode, ddmap, compRequired);
+                    case "group":
```

Comment:
There is a lot of copied code from the previous case. Could that be factored out into methods?

Reply:
I will definitely be coming back to this function later.